File: measure/log.c
#include "log.h"

// void log_performance_results() {

//     bigint *x = NULL;
//     bigint *y = NULL;
//     bigint *z = NULL;

//     int sign1 = rand() % 2;
//     int sign2 = rand() % 2;
//     int wordlen1 = rand() % 96 + 1;
//     int wordlen2 = rand() % 96 + 1;
//     bi_gen_rand(&x, sign1, wordlen1);
//     bi_gen_rand(&y, sign2, wordlen2);
//     FILE *file = fopen("performance_results.csv", "w");
//     fprintf(file, "Algorithm,Execution Time (s),CPU Cycles,Memory Usage (KB)\n");

//     // 예시로 세 개의 곱셈 함수에 대해 성능을 측정
//     fprintf(file, "mul_core_tx,%f,%llu,%zu\n", 
//             measure_execution_time(mul_core_tx, &x, &y, &z),
//             measure_cycles(mul_core_tx, &x, &y, &z),
//             measure_memory_usage(mul_core_tx, &x, &y, &z));

//     fprintf(file, "mul_core_improved,%f,%llu,%zu\n", 
//             measure_execution_time(mul_core_improved, &x, &y, &z),
//             measure_cycles(mul_core_improved, &x, &y, &z),
//             measure_memory_usage(mul_core_improved, &x, &y, &z)
//             //measure_cache_misses(mul_core_improved)
//             );

//     fprintf(file, "mul_core_karatsuba,%f,%llu,%zu\n", 
//             measure_execution_time(mul_core_karatsuba, &x, &y, &z),
//             measure_cycles(mul_core_karatsuba, &x, &y, &z),
//             measure_memory_usage(mul_core_karatsuba, &x, &y, &z)
//             //measure_cache_misses(mul_core_karatsuba)
//             );

//     fclose(file);

//     bi_delete(&x);
//     bi_delete(&y);
//     bi_delete(&z);

// }

File: measure/log.py
import matplotlib.pyplot as plt
import pandas as pd

# CSV 파일에서 성능 결과 읽기
data = pd.read_csv('performance_results.csv')

# 실행 시간 그래프
plt.figure(figsize=(10, 6))
plt.bar(data['Algorithm'], data['Execution Time (s)'], color='b', alpha=0.7)
plt.xlabel('Algorithm')
plt.ylabel('Execution Time (s)')
plt.title('Execution Time Comparison')
plt.show()

# CPU 클럭 사이클 그래프
plt.figure(figsize=(10, 6))
plt.bar(data['Algorithm'], data['CPU Cycles'], color='g', alpha=0.7)
plt.xlabel('Algorithm')
plt.ylabel('CPU Cycles')
plt.title('CPU Cycles Comparison')
plt.show()

# 메모리 사용량 그래프
plt.figure(figsize=(10, 6))
plt.bar(data['Algorithm'], data['Memory Usage (KB)'], color='r', alpha=0.7)
plt.xlabel('Algorithm')
plt.ylabel('Memory Usage (KB)')
plt.title('Memory Usage Comparison')
plt.show()